qos/evm/state_archival.go

```diff
@@ -0,0 +1,190 @@
+//   - Each endpoint is evaluated against this established truth - any endpoint that reports a different balance value is flagged as lacking proper archival data and will be filtered out by QoS.
+type archivalState struct {
+	logger polylog.Logger
+	// archivalCheckConfig contains all configurable values for an EVM archival check.
+	archivalCheckConfig EVMArchivalCheckConfig
+
+	// blockNumberHex is a randomly selected block number from which to check the balance of the contract.
+	//
+	// It is calculated using the `calculateArchivalBlockNumber` method, which selects a block from the range:
+	// 		- earliest possible block = <archivalCheckConfig.ContractStartBlock>
+	// 		- latest possible block = <perceivedBlockNumber> - <archivalCheckConfig.Threshold>
+	//
+	// eg. if perceivedBlockNumber = 100, archivalThreshold = 10, contractStartBlock = 10, then the block number will be between 10 and 90.
+	//
+	// Format will look like: `0x3f8627c`
+	blockNumberHex string
+
+	// balance is the balance of the contract at the block number specified in `blockNumberHex`.
+	// It is determined by reaching a consensus on the balance among `<archivalConsensusThreshold>` endpoints.
+	expectedBalance string
+
+	// balanceConsensus is a map where:
+	//   - key: hex balance value for the archival block number
+	//   - value: number of endpoints that reported the balance
+	//
+	// eg. {"0x1ce31607bc8f16a8c53d80": 5, "0x1ce31607bc8f16a8c53d81": 3}
+	//
+	// When a single hex value with a count of <archivalConsensusThreshold> is reached, the balance
+	// is set as the expected archival balance source of truth for archival validation.
+	balanceConsensus map[string]int
+
+	// TODO_IMPROVE(@commoddity): set an expiry time for the archival state so that a new random block can be selected on an interval.
+}
+
+func (a *archivalState) isEnabled() bool {
+	return a.archivalCheckConfig.Enabled
+}
+
+// processEndpointArchivalData processes all archival-related data from an endpoint
+// and updates the archival state accordingly.
+//
+// Returns true if the endpoint's archival data was processed successfully,
+// false if the endpoint should be skipped (e.g., missing archival balance).
+func (a *archivalState) processEndpointArchivalData(endpoint endpoint) bool {
+	if !a.isEnabled() {
+		return true
+	}
+
+	// Attempt to retrieve the archival balance from the endpoint
+	balance, err := endpoint.getArchivalBalance()
+	if err != nil {
+		return false
+	}
+
+	// Update the consensus map to determine the balance at the perceived block number
+	a.updateConsensusMap(balance)
+	return true
+}
+
+// updateConsensusMap updates the balance consensus map to determine
+// the expected archival balance at the archival block number.
+//
+// Once <archivalConsensusThreshold> endpoints report the same balance,
+// the balance is set as the expected archival balance.
+func (a *archivalState) updateConsensusMap(balance string) {
```

Comment:
`Map` in the method leaks details the caller should not know about. `updateBalanceConsensus` (or similar) may be a better name here.

Reply:
This file has been refactored so this is no longer an issue.